src/dynamic_honey_badger.rs

```diff
@@ -0,0 +1,450 @@
+                    self.start_epoch = hb_batch.epoch + 1;
+                    self.apply_change(change, pub_key_set, sk)?;
+                    batch.change = Some(change.clone());
+                    changed = true;
+                }
+            }
+            // If a new change has a majority, restart DKG.
+            if self.current_majority() != change.as_ref() {
+                if let Some(change) = self.current_majority().cloned() {
+                    self.start_key_gen(change)?;
+                } else {
+                    self.key_gen = None;
+                }
+            }
+            self.output.push_back(batch);
+        }
+        self.messages
+            .extend_with_epoch(self.start_epoch, &mut self.honey_badger);
+        // If `start_epoch` changed, we can now handle some queued messages.
+        if changed {
+            let queue = mem::replace(&mut self.incoming_queue, Vec::new());
+            for (epoch, sender_id, msg) in queue {
+                self.handle_honey_badger_message(&sender_id, epoch, msg)?;
+            }
+        }
+        Ok(())
+    }
+
+    /// Restarts Honey Badger with a new set of nodes, and resets the Key Generation.
+    fn apply_change(
+        &mut self,
+        change: &Change<NodeUid>,
+        pub_key_set: PublicKeySet,
+        sk: ClearOnDrop<Box<SecretKey>>,
+    ) -> Result<()> {
+        self.votes.clear();
+        self.vote_counts.clear();
+        self.key_gen = None;
+        let mut all_uids = self.netinfo.all_uids().clone();
+        if !match *change {
+            Change::Remove(ref id) => all_uids.remove(id),
+            Change::Add(ref id, _) => all_uids.insert(id.clone()),
+        } {
+            debug!("No-op change: {:?}", change);
+        }
+        let netinfo = NetworkInfo::new(self.our_id().clone(), all_uids, sk, pub_key_set);
+        self.netinfo = netinfo.clone();
+        // TODO: Drop the buffer if this node was removed, to become an observer.
+        let buffer = self.honey_badger.drain_buffer();
+        self.honey_badger = HoneyBadger::new(Rc::new(netinfo), self.batch_size, buffer)?;
+        Ok(())
+    }
+
+    /// Starts Key Generation for the set of nodes implied by the `change`.
+    fn start_key_gen(&mut self, change: Change<NodeUid>) -> Result<()> {
+        // Use the existing key shares - with the change applied - as keys for DKG.
+        let mut pub_keys = self.netinfo.public_key_map();
+        if match change {
+            Change::Remove(id) => pub_keys.remove(&id).is_none(),
+            Change::Add(id, pub_key) => pub_keys.insert(id, pub_key).is_some(),
+        } {
+            debug!("No-op change: {:?}", self.current_majority().unwrap());
+        }
+        // TODO: This needs to be the same as `num_faulty` will be in the _new_
+        // `NetworkInfo` if the change goes through. It would be safer to deduplicate.
+        let threshold = (pub_keys.len() - 1) / 3;
+        let sk = self.netinfo.secret_key().clone();
```

Comment:
Hmm... this allocates a secret key on stack. I'm not sure what can be done about it though.

I guess we could change `self.netinfo.secret_key()` to return a pointer to the heap allocated secret-key, and then change `SyncKeyGen.sec_key` to be of type `ClearOnDrop<Box<SecretKey>>`.

Reply:
That's true. I wonder whether the right thing to do would be to turn what's currently `SecretKey` into a private `SecretKeyInner`, and put a `ClearOnDrop<Box<SecretKeyInner>>` inside the new `SecretKey`, so that outside of the `crypto` module there's no way to circumvent it anymore?

Reply:
I created https://github.com/poanetwork/hbbft/issues/89 for it.